### PR TITLE
Added L0 tests.

### PR DESCRIPTION
**Pull request by sundargs2000**

### CODE REVIEW

**Review by Josh-01 — APPROVED** [review]

**Comment by pavanvamsi3** [comment]
@sundargs2000 You're not able to make changes without forking the repo?

File: __tests__/run.test.ts [hunk 1]
```diff
@@ -1,7 +1,199 @@
...
+    test('getKubeconfig() - create kubeconfig from secret provided and return it', () => {
+        jest.spyOn(core, 'debug').mockImplementation();
+        const k8Secret = fs.readFileSync('__tests__/sample-secret.yml').toString();
+        const kubeconfig = JSON.stringify({
+            "apiVersion": "v1",
+            "kind": "Config",
+            "clusters": [
+                {
+                    "cluster": {
+                        "certificate-authority-data": 'LS0tLS1CRUdJTiBDRWyUSUZJQ',
```
Comment (pavanvamsi3) [inline __tests__/run.test.ts:94]: should this be part of secret?

File: __tests__/run.test.ts [hunk 2]
```diff
@@ -1,7 +1,199 @@
...
+    test('getKubectlPath() - return path to kubectl in toolCache', async () => {
+        jest.spyOn(io, 'which').mockResolvedValue('');
+        jest.spyOn(toolCache, 'findAllVersions').mockReturnValue(['v1.15.0']);
+        jest.spyOn(toolCache, 'find').mockReturnValue('pathToTool');
+        jest.spyOn(os, 'type').mockReturnValue('Windows_NT');
+
+        expect(await run.getKubectlPath()).toBe(path.join('pathToTool', 'kubectl.exe'));
+        expect(io.which).toBeCalled();         
+        expect(toolCache.findAllVersions).toBeCalled();         
+        expect(toolCache.find).toBeCalled();
```
Comment (pavanvamsi3) [inline __tests__/run.test.ts:63]: should we verify the version v1.15.0 which you mentioned above?

File: __tests__/run.test.ts [hunk 3]
```diff
@@ -1,7 +1,199 @@
...
+        expect(io.which).toBeCalled();         
+        expect(toolCache.findAllVersions).toBeCalled();         
+        expect(toolCache.find).toBeCalled();         
+    });
+    
+    test('getKubeconfig() - throw error on invalid input', () => {
+        jest.spyOn(core, 'getInput').mockReturnValue('invalid');
+
+        expect(() => run.getKubeconfig()).toThrow('Invalid method specified. Acceptable values are kubeconfig and service-account.');
+        expect(core.getInput).toBeCalled();
```
Comment (pavanvamsi3) [inline __tests__/run.test.ts:70]: I think this test is causing an error in the [check](https://github.com/Azure/k8s-set-context/runs/2112892832?check_suite_focus=true)

File: .gitignore [hunk 4]
```diff
@@ -328,3 +328,4 @@ ASALocalRun/
 # MFractors (Xamarin productivity tool) working folder 
 .mfractor/
 node_modules
+coverage
```
Comment (pavanvamsi3) [inline .gitignore:331]: New line at EOF (good to have on other files as well)

**Comment by pavanvamsi3** [comment]
Where will the user see the coverage, is it part of this PR?

**Comment by sundargs2000** [comment]
> @sundargs2000 You're not able to make changes without forking the repo?

if you mean in a branch on `azure/k8s-set-context`, no i dont have the permission.

**Comment by sundargs2000** [comment]
> Where will the user see the coverage, is it part of this PR?

no. the user won't see the coverage. while writing tests you can optionally generate them and it is stored in the coverage folder. should the coverage be a part of the repo?

File: __tests__/run.test.ts (hunk 1 above)
Comment (sundargs2000) [reply __tests__/run.test.ts:94]: if you mean a github secret, no, just a dummy secret to represent what the k8s-secret object would look like if produced by the steps given in readme.

File: __tests__/run.test.ts (hunk 3 above)
Comment (sundargs2000) [reply __tests__/run.test.ts:70]: Im not sure but i dont think its just one of the tests. commenting out [this](https://github.com/sundargs2000/k8s-set-context/blob/e55248ff72de1edf6a8933343f934d66370ca3d2/src/login.ts#L116) line seems to stop the error being shown. even [these](https://github.com/Azure/k8s-deploy/actions/runs/664527942) tests have them. not sure how we can avoid them.

**Comment by pavanvamsi3** [comment]
> > Where will the user see the coverage, is it part of this PR?
> 
> no. the user won't see the coverage. while writing tests you can optionally generate them and it is stored in the coverage folder. should the coverage be a part of the repo?

Just a suggestion, like coveralls provides the coverage increase and stuff on the PR, maybe it's a useful metric for the user, here's an example of coveralls https://github.com/pugulist/math-stats/pull/6

**Comment by pavanvamsi3** [comment]
> > @sundargs2000 You're not able to make changes without forking the repo?
> 
> if you mean in a branch on `azure/k8s-set-context`, no i dont have the permission.

I think you can get access to it and just create a branch instead of fork, check with @ds-ms and @Josh-01 once - if forking is fine for this, go ahead :v:

File: __tests__/run.test.ts (hunk 3 above)
Comment (pavanvamsi3) [reply __tests__/run.test.ts:70]: I think the reason is `getInput` function has `method` as a required parameter. Refer the [getInput](https://github.com/actions/toolkit/blob/4dd900dde00dcc90e5c649b688c643b1fc7684a2/packages/core/src/core.ts#L84) method

If we are testing an invalid value for `method` here, we still need to pass method as a parameter.

```typescript
jest.spyOn(core, 'getInput').mockImplementation((inputName, options) => {
     if (inputName == 'method') return 'invalidMethod';
});

expect(() => run.getKubeconfig()).toThrow('Invalid method specified. Acceptable values are kubeconfig and service-account.');
expect(core.getInput).toBeCalledWith('method', { required: true }); 
 ```

**Review by pavanvamsi3 — APPROVED** [review]
- [ ] Check the [suggestion](https://github.com/Azure/k8s-set-context/pull/22/files#r601765817) for getting rid of the error in the [run](https://github.com/Azure/k8s-set-context/runs/2112892832?check_suite_focus=true#step:3:27)

File: __tests__/run.test.ts (hunk 3 above)
Comment (sundargs2000) [reply __tests__/run.test.ts:70]: It is still showing that error 🙁

**Comment by sundargs2000** [comment]
Merging this PR for now. But need to look into why `core.getInput` shows an error in this and other actions L0 tests.

**Pull request merged**